### ISSUE TITLE
Is ASPT disabled for SM951/PM951 upstream?

### ISSUE THREAD

**Issue by davidgfnet** [damige/linux-nvme#26]
Does the kernel 4.11 and onwards have the default quirk to disable low power states for SM951/PM951?
Does this quirk disable ALL power savings or just some? I do have that HDD and never had any probem with the HDD using your patches.

**Comment by damige** [comment]
My understanding is: deepest sleep has been disabled on samsung nvme drives.
And toshiba drives have APST disabled completely

Source: https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1678184/comments/43


https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1678184/comments/44

**Comment by davidgfnet** [comment]
Ok I see. But on 4.10 with the APST patch the deepest sleep is NOT disabled right?

**Comment by damige** [comment]
Correct, it *should* unconditionally enable it for nvme devices

**Comment by davidgfnet** [comment]
I see. So two more questions, please:

-  Do you have any problems with the lowest sleep level? Cause I've never had :D
- Why they decided to go that direction? Maybe it's just a batch of drives with bad firmware. In that case, can the quirk be enabled/disabled through /sys or something similar? Or maybe a kernel arg? I dont wanna updatre to 4.11 just because of that. Either that or I will create a rollback patch for 4.11 and apply it in my copr builds

many thanks!

**Comment by damige** [comment]
1. I have had some issues a while back, but they were quite rare. Since running the new bios for my xps13 9350 i have not seen it happen anymore. (fingers crossed)

2. From my understanding they are working with dell/lenovo to get the problems fixed. Until it is, they have put quirks in place to make the kernel stable. I am not aware of any overrides on the quirk. If you find something it would be nice to know!

**Comment by davidgfnet** [comment]
Cool thanks for the feedback. It might depend on BIOS versions as you say,
there's lots of crap in the firmware these days. Of course HW guys will
blame FW and the FW/OS guys will blame the HW, what a classic move :)

There's a new BIOS from the 2nd of May, have you tried it yet?

**Comment by davidgfnet** [comment]
Just took a look at 4.11, the patch doesnt affect XPS 9350:

```
static unsigned long check_dell_samsung_bug(struct pci_dev *pdev)
{
        if (pdev->vendor == 0x144d && pdev->device == 0xa802) {
                /*
                 * Several Samsung devices seem to drop off the PCIe bus
                 * randomly when APST is on and uses the deepest sleep state.
                 * This has been observed on a Samsung "SM951 NVMe SAMSUNG
                 * 256GB", a "PM951 NVMe SAMSUNG 512GB", and a "Samsung SSD
                 * 950 PRO 256GB", but it seems to be restricted to two Dell
                 * laptops.
                 */
                if (dmi_match(DMI_SYS_VENDOR, "Dell Inc.") &&
                    (dmi_match(DMI_PRODUCT_NAME, "XPS 15 9550") ||
                     dmi_match(DMI_PRODUCT_NAME, "Precision 5510")))
                        return NVME_QUIRK_NO_DEEPEST_PS;
        }

        return 0;
}
```

Also the quirk you describe only affects Toshiba:

```
static const struct nvme_core_quirk_entry core_quirks[] = {
        {
                /*
                 * This Toshiba device seems to die using any APST states.  See:
                 * https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1678184/comments/11
                 */
                .vid = 0x1179,
                .mn = "THNSF5256GPUK TOSHIBA",
                .quirks = NVME_QUIRK_NO_APST,
        }
};
```

So I conclude 4.11 should be allright for us. Will try to install it and see how it works.

**Comment by tlvince** [comment]
Yeah, currently it's only disabled on specific Dell laptops: https://github.com/torvalds/linux/commit/ff5350a86b20de23991e474e006e2ff2732b218e
